casper/src/main/scala/coop/rchain/casper/merging/MergeScope.scala

```diff
@@ -0,0 +1,195 @@
+import coop.rchain.sdk.dag.merging.ConflictResolutionLogic.{conflictScope, finalScope}
+import coop.rchain.sdk.syntax.all._
+import coop.rchain.shared.{Log, Stopwatch}
+import scodec.bits.ByteVector
+
+/**
+  * The scope to merge.
+  * @param finalScope     blocks that cannot be altered
+  * @param conflictScope  blocks that can be altered
+  * @param baseState      base state to merge into
+  */
+final case class MergeScope(
+    finalScope: Set[BlockHash],
+    conflictScope: Set[BlockHash],
+    baseState: Blake2b256Hash
+)
+
+object MergeScope {
+
+  private def findGenesis(
+      dagData: Map[BlockHash, Message[BlockHash, Validator]]
+  ): Option[BlockHash] = {
+    val candidates = dagData.filter {
+      case (_, m) =>
+        val edge      = m.seen == Set(m.id)
+        val seenByAll = dagData.valuesIterator.forall(_.seen.contains(m.id))
+        edge && seenByAll
+    }
+    if (candidates.size > 1) none[BlockHash] else candidates.headOption.map(_._1)
+  }
+
+  /**
+    * Use dag data to optimise the merge scope.
+    * For now covers only the case of genesis ceremony, when final set is empty so
+    * genesis block can be removed from conflict set and merge can be done into genesis post state.
+    */
+  def optimise[F[_]: Applicative](
+      mergeScope: MergeScope,
+      dagData: Map[BlockHash, Message[BlockHash, Validator]],
+      postState: BlockHash => F[Blake2b256Hash]
+  ): F[MergeScope] = {
+    val MergeScope(fScope, cScope, _) = mergeScope
+    val genesisCase = {
+      val gOpt   = findGenesis(dagData)
+      val errStr = "Cannot create merge scope. Genesis case but genesis block cannot be found"
+      assert(gOpt.isDefined, errStr)
+      val genesis = gOpt.get
+      postState(genesis).map(MergeScope(fScope, cScope - genesis, _))
+    }
+    // if final scope is empty - do special case for genesis
+    if (fScope.isEmpty) genesisCase
+    else mergeScope.pure
+  }
+
+  def apply(
+      mergeFringe: Set[BlockHash],
+      finalFringe: Set[BlockHash],
+      finalState: Blake2b256Hash,
+      dagData: Map[BlockHash, Message[BlockHash, Validator]]
+  ): MergeScope = {
+    val seen      = dagData.getUnsafe(_: BlockHash).seen
+    val cScope    = conflictScope(mergeFringe, finalFringe, seen)
+    val finalizer = Finalizer(dagData)
+    val lFringe   = finalizer.lowestFringe(cScope.map(dagData)).map(_.id)
+    val fScope    = finalScope(finalFringe, lFringe, seen)
+    MergeScope(fScope, cScope, finalState)
+  }
+
+  def merge[F[_]: Concurrent: Log](
+      mergeScope: MergeScope,
+      fringeStates: Map[Set[BlockHash], FringeData],
+      historyRepository: RhoHistoryRepository[F],
+      blockIndex: BlockHash => F[BlockIndex],
+      rejectionCost: DeployChainIndex => Long = DeployChainIndex.deployChainCost
+  ): F[(Blake2b256Hash, Set[ByteString])] = {
+    val MergeScope(finalScope, conflictScope, baseState) = mergeScope
+
+    // if some indices can be computed - merge is impossible.
+    val loadIndices =
```

Comment:
This can be used to check when merge is possible, which is one requirements for declaration of bootstrap procedure successfully accomplished.